cpp/src/arrow/compute/expression.cc

```diff
@@ -1242,8 +1273,92 @@ struct Inequality {
                             /*insert_implicit_casts=*/false, &exec_context);
   }
 
+  /// Simplify an `is_in` value set against an inequality guarantee.
+  ///
+  /// Simplifying an `is_in` predicate involves filtering out any values from
+  /// the value set that cannot possibly be found given the guarantee. For
+  /// example, if we have the predicate 'x is_in [1, 2, 3, 4]' and the guarantee
+  /// 'x > 2', then the simplified predicate 'x is_in [3, 4]' is equivalent.
+  /// This can be done efficiently if the value set is sorted and unique by
+  /// binary searching the inequality gound and slicing the value set
+  /// accordingly.
+  ///
+  /// \pre `guarantee` is non-nullable
+  /// \pre `guarantee.bound` is a scalar
+  /// \pre `guarantee.bound.type()->id() == value_set->type_id()`
+  /// \pre `value_set` is non-empty
+  /// \return a simplified value set, or a bool if the simplification results in
+  ///   a boolean literal predicate.
+  template <typename ArrowType>
+  static std::variant<std::shared_ptr<Array>, bool> SimplifyIsInValueSet(
```

Comment:
it almost feels like this binary search/lower_bound functionality should be its own compute function. though idk if this is overkill for what we're trying to do.

Reply:
Totally overkill. The binary search should be extracted as a function here because it really is the only part that needs ~17 template specializations. That will require moving the `switch` into this function.

Reply:
I don't follow the part about moving the `switch` into this function.

I'm mainly unsure of whether we should handle the case where the guarantee type and value set type are not identical. Previously, the comparison compute function would have handled this for us by dispatching on the best kernel, but currently the code will just abort simplification if this is the case. Any thoughts on whether we should handle this here and the best way to go about this?

Reply:
You can return a value that indicates the simplification should be aborted.

This is a common difficulty in kernel development: specializing by type (using templates) only the kernel loops (in this case, the loop is the binary-search) so that binary size doesn't explode.

<!-- Other libraries don't care so much about binary size, but Arrow's success as a basic dependency on so many libraries comes from not being a multi-GB library like many template-heave compute kernel libraries can be. -->

Reply:
I took a read through the comparison kernel code and realized that supporting inexact dispatch here isn't so bad.